### ISSUE TITLE
Promises/coroutines

### ISSUE THREAD

**Issue by mmasiukevich** [vimeo/psalm#1283]
Greetings

I would like to discuss how [promises](https://github.com/amphp/amp/blob/master/lib/Promise.php) should be described correctly (in theory).
For example,
https://paste.ofcode.org/YqEeBz3qdVsxHzjm8XkGxu
Such a description is expected to lead to an error

Yes, I understand that I can specify the type myself

/** @var string $myString */
 $myString = yield qwerty();

but this greatly undermines the value of the analysis.

Is there any plan to support promises (coroutine)?

**Comment by muglug** [comment]
Not going to be supported natively because they're vendor-dependent, but you can stub them out:

`amp_stubs.php`:
```php
<?php

namespace Amp;

/**
 * @template TReturn
 * @param callable():\Generator<mixed, mixed, mixed, TReturn> $gen
 * @return Promise<TReturn>
 */
function coroutine(callable $gen) : Promise {}

/**
 * @template TReturn
 * @param callable():(\Generator<mixed, mixed, mixed, TReturn>|null) $gen
 * @return Promise<TReturn>
 */
function call(callable $gen) : Promise {}

/**
 * @template TReturn
 */
class Promise {
    /**
     * @return TReturn
     */
    function wait() {}
}
```

If that works for you I can create an appropriate Psalm plugin

**Comment by mmasiukevich** [comment]
It would be nice that amphp also supported @template

**Comment by muglug** [comment]
You could always PR those additions (but make sure to use `@psalm-param`, `@psalm-return`)

cc @kelunik

**Comment by kelunik** [comment]
The main issue here is not `call` or `coroutine` using `@template`, but rather yielding promises in generators, which injects the resolved value into the generator again, so it needs special support.

**Comment by kelunik** [comment]
`yield`ing a promise unwraps it, so the result type of the `yield` expression must be inferred from the yielded promise. If that doesn't answer your question, then I didn't get what you mean, could you rephrase it?

**Comment by muglug** [comment]
> `yield`ing a promise unwraps it

Where does that magic happen?

**Comment by muglug** [comment]
Ok, I think I understand it.

`call\asynCall` take as commands a callable that can return a generator. That generator is wrapped by an object that injects values when `yield`ing Promises, so that the actual promise is not yielded - rather the value inside. I was confused until I realised the generator was always wrapped.

**Comment by kelunik** [comment]
@muglug Correct, the class that's responsible for that is `Coroutine`.

**Comment by trowski** [comment]
@muglug I just noticed that `coroutine()` in your stub is wrong. In Amp, that function takes a callable returning a Generator and returns a callable that returns a Promise.

```php
/**
 * @template TReturn
 * @param callable():\Generator<mixed, mixed, mixed, TReturn> $gen
 * @return callable():Promise<TReturn>
 */
function coroutine(callable $gen) : callable {}

**Comment by muglug** [comment]
@trowski Thanks, I've fixed that. The way I got promises doing the right thing was to use the `Generator<TKey, TValue, TSend, TReturn>` `TSend` type parameter such that any `yield` commands below it return that value. I need to write a proper test for it, though.

**Comment by muglug** [comment]
This is more-or-less done

**Comment by kelunik** [comment]
@muglug Should the stubs for Amp be removed now that we have them in the code?

**Comment by muglug** [comment]
They already have been!
